infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java

```diff
@@ -139,6 +128,54 @@ protected void internalStart(Map<String, String> startOptions) throws Infrastruc
     }
   }
 
+  /**
+   * Bootstraps machines in parallel.
+   *
+   * @throws InfrastructureException when any error occurs while bootstrapping machines
+   * @throws InterruptedException when machines bootstrapping was interrupted
+   */
+  private void bootstrapMachines() throws InfrastructureException, InterruptedException {
+    CompletableFuture<InfrastructureException> firstFailed = new CompletableFuture<>();
+    List<CompletableFuture> checkMachines = new ArrayList<>(machines.size());
+    for (OpenShiftMachine machine : machines.values()) {
+      checkMachines.add(
+          CompletableFuture.runAsync(
+                  () -> {
+                    try {
+                      machine.waitRunning(machineStartTimeoutMin);
+                      bootstrapMachine(machine);
+                      checkMachineServers(machine);
+                    } catch (InfrastructureException ex) {
+                      sendFailedEvent(machine.getName(), ex.getMessage());
+                      firstFailed.completeExceptionally(ex);
+                      throw new CompletionException(ex);
+                    } catch (InterruptedException ei) {
+                      throw new CompletionException(ei);
+                    }
+                  })
+              .thenAccept(
+                  a -> {
+                    if (!firstFailed.isCompletedExceptionally()) {
+                      sendRunningEvent(machine.getName());
+                    }
+                  }));
+    }
+
+    CompletableFuture<Void> allAvailable =
+        CompletableFuture.allOf(checkMachines.toArray(new CompletableFuture[0]));
+    try {
+      CompletableFuture.anyOf(allAvailable, firstFailed).join();
+    } catch (CompletionException e) {
+      if (firstFailed.isDone()) {
+        throw firstFailed.join();
```

Comment:
You expect here get InfrastructureException but you won't get it here. Because this completable future is complete exceptionally [here](https://github.com/eclipse/che/pull/7379/files#diff-fd5e0f37b880417aed392d5dc7020dd6R150).
And then as far as I understand `firstFailed.join();` will throw CompletionException caused by InfrastructureException.